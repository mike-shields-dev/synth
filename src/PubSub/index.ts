export { ControlChangeSubscriber, publishControlChange } from './ControlChange';
export type { ControlChange } from './ControlChange/types';
export { MidiNoteOnSubscriber, publishMidiNoteOn } from './MidiNoteOn';
export type { MidiNoteOn } from './MidiNoteOn/types';
export { NoteOffSubscriber, publishNoteOff } from './NoteOff';
export type { NoteOff } from './NoteOff/types';

