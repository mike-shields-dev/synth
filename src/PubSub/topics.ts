const
    MIDI = 'midi',
    MIDI_CC = 'midi.controlChange',
    MIDI_NOTE_ON = 'midi.note.on',
    MIDI_NOTE_OFF = 'midi.note.off',
    UI_CC = 'ui.controlChange',
    FOCUS = 'focus';

export {
    MIDI, 
    MIDI_CC, 
    MIDI_NOTE_ON,
    MIDI_NOTE_OFF,
    UI_CC,
    FOCUS,
}