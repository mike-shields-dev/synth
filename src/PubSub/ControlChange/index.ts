import PubSub from 'pubsub-js';
import { CONTROL } from '../topics';
import { ControlChange, ControlChangeHandler } from './types';

function publishControlChange(data: ControlChange) {
  PubSub.publish(CONTROL, data);
}

class ControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: ControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      CONTROL,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishControlChange,
  ControlChangeSubscriber
};

