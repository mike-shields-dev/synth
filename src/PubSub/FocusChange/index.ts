import PubSub from "pubsub-js";
import { FOCUS } from '../topics';
import { FocusChangeHandler } from "./types";

function publishFocusChange(data: string) {
    PubSub.publish(FOCUS, data);
}

class FocusChangeSubscriber {
    private subscription: string;

    constructor(focusHandler: FocusChangeHandler) {
        this.subscription = PubSub.subscribe(
            FOCUS,
            focusHandler,
        );
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    }
}

export {
    publishFocusChange,
    FocusChangeSubscriber
};
