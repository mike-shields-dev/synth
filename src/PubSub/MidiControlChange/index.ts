import PubSub from 'pubsub-js';
import { MIDI_CC } from '../topics';
import { MidiControlChange, MidiControlChangeHandler } from './types';

function publishMidiControlChange(data: MidiControlChange) {
  PubSub.publish(MIDI_CC, data);
}

class MidiControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: MidiControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      MIDI_CC,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishMidiControlChange,
  MidiControlChangeSubscriber
};

