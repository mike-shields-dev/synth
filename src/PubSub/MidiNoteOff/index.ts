import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { MidiNoteOff, MidiNoteOffHandler } from './types';

function publishMidiNoteOff(payload: MidiNoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, payload)
}

class MidiNoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: MidiNoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOff, MidiNoteOffSubscriber };
