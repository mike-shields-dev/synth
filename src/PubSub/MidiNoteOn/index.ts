import PubSub from "pubsub-js";
import { MIDI_NOTE_ON } from "../topics";
import { MidiNoteOn, MidiNoteOnHandler } from "./types";

function publishMidiNoteOn(payload: MidiNoteOn) {
    PubSub.publish(MIDI_NOTE_ON, payload)
}

class MidiNoteOnSubscriber {
    private subscription: any;

    constructor(noteOnHandler: MidiNoteOnHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_ON,
            noteOnHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOn, MidiNoteOnSubscriber };
