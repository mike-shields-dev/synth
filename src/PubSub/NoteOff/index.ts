import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { NoteOff, NoteOffHandler } from './types';

function publishNoteOff(data: NoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, data)
}

class NoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: NoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishNoteOff, NoteOffSubscriber };
