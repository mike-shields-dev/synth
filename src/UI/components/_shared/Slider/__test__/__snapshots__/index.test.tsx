import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { vi } from 'vitest';
import { Slider } from '../..';
import { CONTROL } from '../../../../../../PubSub/topics';

const validProps = {
    controlChangeNumber: 70,
    group: 'groupName',
    initVal: 63,
    isFocused: true,
    parameter: 'parameter',
    scalers: {
        in: vi.fn((n) => n / 2),
        out: vi.fn((n) => n * 2),
    }
}

describe('Slider', () => {
    it('matches snapshot', () => {
        const { container } = render(<Slider {...validProps} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    beforeEach(() => render(< Slider {...validProps} />));
    afterEach(() => vi.resetAllMocks());

    it('renders a Slider component', () => {
        expect(screen.getByRole('slider', { name: 'Parameter' })).toBeInTheDocument();
    });

    it('renders a single range input', () => {
        expect(screen.getAllByRole('slider')).toHaveLength(1);
    });

    it('renders a range input with a min value of 0', () => {
        expect(screen.getByRole('slider')
            .getAttribute('min'))
            .toEqual('0');
    });

    it('renders a range input with a max value of 127', () => {
        expect(screen.getByRole('slider')
            .getAttribute('max'))
            .toEqual('127');
    });

    it('renders a range input with the correctly scaled value', () => {
        const slider = screen.getByRole('slider', { name: 'Parameter' });
        const scaledValue = validProps.scalers.in(validProps.initVal)

        expect(slider).toHaveValue(`${scaledValue}`);
    });

    it('renders an output element', () => {
        expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('renders an output element associated with the input element', ()=> {
        expect(screen.getByRole('status')
            .getAttribute('for'))
            .toBe(`${validProps.group}-${validProps.parameter}`);
    });

    it('renders an output element that has a scaled value', async () => {
        const slider = screen.getByRole('slider', { name: 'Parameter' });
        const newValue = 23;
        const scaledValue = validProps.scalers.out(newValue)

        fireEvent.change(slider, { target: { value: newValue } });

        expect(await screen.findByText(scaledValue)).toBeInTheDocument();
    });

    describe("when provided isFocused prop is equal to true", () => {
        it('publishes a UI Control Change message when the user moves the slider', async () => {
            const slider = screen.getByRole('slider', { name: 'Parameter' });
            const newValue = 64;
            const spy = vi.fn();

            PubSub.subscribe(CONTROL, spy);
            
            fireEvent.change(slider, { target: { value: newValue } });
    
            await waitFor(() => {
                expect(spy).toHaveBeenCalled()
            });
        });
    
        it('updates the slider value when a MIDI control change is received', async () => {
            const newValue = 88;
    
            PubSub.publish(CONTROL, {
                controlChangeNumber: validProps.controlChangeNumber,
                value: newValue,
            });
    
            await waitFor(() => {
                expect(screen.getByRole('slider')).not.toHaveValue(`${newValue}`);
            });
        });
    });

    describe("when provided isFocused prop is equal to false", () => {
        it('does not updates the slider value when a MIDI control change is received', async () => {
            validProps.isFocused = false;
            const initValue = `${validProps.scalers.in(validProps.initVal)}`;
            const newValue = 88;
    
            PubSub.publish(CONTROL, {
                controlChangeNumber: validProps.controlChangeNumber,
                value: newValue,
            });
    
            await waitFor(() => {
                expect(screen.getByRole('slider')).toHaveValue(initValue);
            });
        });
    });
});