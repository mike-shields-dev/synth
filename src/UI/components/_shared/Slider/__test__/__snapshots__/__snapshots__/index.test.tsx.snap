// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider > matches snapshot 1`] = `
<div>
  <label
    for="groupName-parameter"
  >
    Parameter
  </label>
  <input
    id="groupName-parameter"
    max="127"
    min="0"
    step="0.05"
    type="range"
    value="31.5"
  />
  <output
    aria-label="Group Name parameter"
    for="groupName-parameter"
  >
    63
  </output>
</div>
`;
